### ISSUE TITLE
SIGSEGV in a thread hangs the mutator

### ISSUE THREAD

**Issue by cuviper** [dyninst/dyninst#100]
This was a side issue noted from #98.  A `SIGSEGV` in one thread will kill the whole process, but Dyninst hangs.  Consider this program as a mutatee:

``` c
#include <pthread.h>
#include <signal.h>

void *
segv_thread(void *arg)
{
    raise(SIGSEGV);
    return NULL;
}

int
main()
{
    pthread_t thread;
    return pthread_create(&thread, NULL, segv_thread, NULL)
            || pthread_join(thread, NULL);
}
```

We can try to run through `parseThat`, and it hangs:

```
$ timeout -s INT 30  parseThat ./segv-thread
[ Processing ./segv-thread ] ==================================================
Creating new BPatch object.... done.
Requesting notification of mutatee exit.... done.
Requestion notification of mutator fork.... done.
Forking mutatee process.... Mutatee launched as process 20992.
Retrieving BPatch_image object.... done.
Parsing image for module information.... Found 356 module(s).
Continuing execution of mutatee.... done.
Waiting for child termination....
    Checking child termination status.... done.
    Checking child termination status.... done.
    Checking child termination status.... done.

Monitor received SIGINT.  Requesting mutator to exit...
parseThat: /home/jistone/paradyn/dyninst/proccontrol/src/process.C:979: static bool int_process::waitAndHandleEvents(bool): Assertion `!recurse' failed.

*
* Abnormal case: Dyninst mutator terminated via signal 6.  Monitor exiting.
* Core file generated.
*
Allowing mutatee (20992) time to exit gracefully...
Times up.  Sending SIGKILL to 20992.
```

If I move the segfault into the main thread, it cleans up properly:

```
$ timeout -s INT 30  parseThat ./segv-thread
[ Processing ./segv-thread ] ==================================================
Creating new BPatch object.... done.
Requesting notification of mutatee exit.... done.
Requestion notification of mutator fork.... done.
Forking mutatee process.... Mutatee launched as process 17320.
Retrieving BPatch_image object.... done.
Parsing image for module information.... Found 356 module(s).
Continuing execution of mutatee.... done.
Waiting for child termination....
    Checking child termination status.... done.
    Checking child termination status.... done.
    Checking child termination status.... done.
    Mutatee exited normally and returned 0
    Mutatee exited via signal 11
Dyninst mutator exited normally and returned 255.
[ Done processing ./segv-thread ] ---------------------------------------------
Analysis complete.
```

**Comment by wrwilliams** [comment]
Note that parseThat is hanging because its mutator process (don't ask why it spawns a process for the mutator, I don't know) is asserting (signal 6). So if we can get at the assert, that should shed light on things.

**Comment by cuviper** [comment]
True, although that assert doesn't show until after the timeout SIGINT is delivered.

**Comment by wrwilliams** [comment]
...why *is* our handler thread trying to recurse, though?

**Comment by wrwilliams** [comment]
Okay, parseThat -S is far more useful here; it gives us the view directly from the mutator. What I'm seeing is that the ptracer thread is still around, the generator and handler have exited, and nobody's actually kicked the main thread out of its waitAndHandleEvents call yet.

**Comment by wrwilliams** [comment]
Our handler thread and generator thread both vanish quietly along paths that shouldn't.

**Comment by wrwilliams** [comment]
Understood; I'm just trying to separate the oddities from the real bugs.

I'm not seeing actual thread exits for handler/generator under strace, but gdb isn't picking up the threads when we break in during the hang. Which means either my interpretation of the data is bad, or something really weird (and probably should-be-impossible) is going on.

Mind trying with -S under systemtap so we can see mutator-side thread creation and destruction without the noise of strace?

**Comment by cuviper** [comment]
Here's a pretty thorough trace, hope it's not too much.  I've included `kprocess` events (~= user threads), signals, waits, and `PTRACE_CONT`.

```
  6895   6895        parseThat kprocess.exec_complete    
  6895   6896        parseThat kprocess.start            
  6895   6897        parseThat kprocess.start            
  6895   6898        parseThat kprocess.start            
  6899   6899        parseThat kprocess.start            
  6899   6899        parseThat kprocess.exec             
  6895   6898        parseThat syscall.wait4             -1, 0x7fdac3ffed1c, __WALL, 0x0
  6899   6899      segv-thread signal.send               SIGTRAP -> 6899 6899 segv-thread
  6899   6899      segv-thread kprocess.exec_complete    
  6899   6899      segv-thread signal.send               SIGCHLD -> 6895 6897 parseThat
  6895   6898        parseThat syscall.wait4.return      6899 (WSTOPSIG=SIGTRAP)
  6895   6897        parseThat syscall.ptrace            PTRACE_CONT, 6899, 0x0, SIG_0
  6895   6897        parseThat syscall.ptrace.return     0
  6899   6899      segv-thread signal.send               SIGTRAP -> 6899 6899 segv-thread
  6899   6899      segv-thread signal.send               SIGCHLD -> 6895 6897 parseThat
  6895   6898        parseThat syscall.wait4             -1, 0x7fdac3ffed1c, __WALL, 0x0
  6895   6898        parseThat syscall.wait4.return      6899 (WSTOPSIG=SIGTRAP)
  6899   6899      segv-thread signal.send               SIGTRAP -> 6899 6899 segv-thread
  6899   6899      segv-thread signal.send               SIGCHLD -> 6895 6897 parseThat
  6895   6898        parseThat syscall.wait4             -1, 0x7fdac3ffed1c, __WALL, 0x0
  6895   6898        parseThat syscall.wait4.return      6899 (WSTOPSIG=SIGTRAP)
  6895   6897        parseThat syscall.ptrace            PTRACE_CONT, 6899, 0x0, SIG_0
  6895   6897        parseThat syscall.ptrace.return     0
  6895   6898        parseThat syscall.wait4             -1, 0x7fdac3ffed1c, __WALL, 0x0
  6899   6899      segv-thread signal.send               SIGTRAP -> 6899 6899 segv-thread
  6899   6899      segv-thread signal.send               SIGCHLD -> 6895 6897 parseThat
  6895   6898        parseThat syscall.wait4.return      6899 (WSTOPSIG=SIGTRAP)
  6895   6898        parseThat syscall.wait4             -1, 0x7fdac3ffed1c, __WALL, 0x0
  6899   6899      segv-thread signal.send               SIGTRAP -> 6899 6899 segv-thread
  6899   6899      segv-thread signal.send               SIGCHLD -> 6895 6897 parseThat
  6895   6898        parseThat syscall.wait4.return      6899 (WSTOPSIG=SIGTRAP)
  6895   6897        parseThat syscall.ptrace            PTRACE_CONT, 6899, 0x0, SIG_0
  6895   6897        parseThat syscall.ptrace.return     0
  6895   6898        parseThat syscall.wait4             -1, 0x7fdac3ffed1c, __WALL, 0x0
  6899   6899      segv-thread signal.send               SIGTRAP -> 6899 6899 segv-thread
  6899   6899      segv-thread signal.send               SIGCHLD -> 6895 6897 parseThat
  6895   6898        parseThat syscall.wait4.return      6899 (WSTOPSIG=SIGTRAP)
  6895   6897        parseThat syscall.ptrace            PTRACE_CONT, 6899, 0x0, SIG_0
  6895   6897        parseThat syscall.ptrace.return     0
  6895   6898        parseThat syscall.wait4             -1, 0x7fdac3ffed1c, __WALL, 0x0
  6899   6901      segv-thread kprocess.start            
  6899   6899      segv-thread signal.send               SIGCHLD -> 6895 6897 parseThat
  6899   6901      segv-thread signal.send               SIGCHLD -> 6895 6897 parseThat
  6895   6898        parseThat syscall.wait4.return      6899 (WSTOPSIG=SIGTRAP | PTRACE_EVENT_CLONE << 8)
  6899   6899      segv-thread signal.send               SIGCHLD -> 6895 6897 parseThat
  6895   6898        parseThat syscall.wait4             -1, 0x7fdac3ffed1c, __WALL, 0x0
  6895   6898        parseThat syscall.wait4.return      6899 (WSTOPSIG=TRACESYSGOOD)
  6895   6898        parseThat syscall.wait4             -1, 0x7fdac3ffed1c, __WALL, 0x0
  6895   6898        parseThat syscall.wait4.return      6901 (WSTOPSIG=SIGSTOP)
  6895   6897        parseThat syscall.ptrace            PTRACE_CONT, 6899, 0x0, SIG_0
  6895   6897        parseThat syscall.ptrace.return     0
  6899   6899      segv-thread signal.send               SIGTRAP -> 6899 6899 segv-thread
  6895   6897        parseThat syscall.ptrace            PTRACE_CONT, 6901, 0x0, SIG_0
  6899   6899      segv-thread signal.send               SIGCHLD -> 6895 6897 parseThat
  6895   6897        parseThat syscall.ptrace.return     0
  6895   6898        parseThat syscall.wait4             -1, 0x7fdac3ffed1c, __WALL, 0x0
  6895   6898        parseThat syscall.wait4.return      6899 (WSTOPSIG=SIGTRAP)
  6895   6895        parseThat signal.send               SIGSTOP -> 6899 6899 segv-thread
  6895   6898        parseThat syscall.wait4             -1, 0x7fdac3ffed1c, __WALL, 0x0
  6895   6895        parseThat signal.send               SIGSTOP -> 6899 6901 segv-thread
  6899   6901      segv-thread signal.send               SIGCHLD -> 6895 6897 parseThat
  6895   6897        parseThat syscall.ptrace            PTRACE_CONT, 6899, 0x0, SIG_0
  6895   6897        parseThat syscall.ptrace.return     0
  6895   6898        parseThat syscall.wait4.return      6901 (WSTOPSIG=SIGSTOP)
  6899   6899      segv-thread signal.send               SIGCHLD -> 6895 6897 parseThat
  6895   6898        parseThat syscall.wait4             -1, 0x7fdac3ffed1c, __WALL, 0x0
  6895   6898        parseThat syscall.wait4.return      6899 (WSTOPSIG=SIGSTOP)
  6899   6899      segv-thread signal.send               SIGTRAP -> 6899 6899 segv-thread
  6899   6899      segv-thread signal.send               SIGCHLD -> 6895 6897 parseThat
  6895   6898        parseThat syscall.wait4             -1, 0x7fdac3ffed1c, __WALL, 0x0
  6895   6898        parseThat syscall.wait4.return      6899 (WSTOPSIG=SIGTRAP)
  6895   6897        parseThat syscall.ptrace            PTRACE_CONT, 6899, 0x0, SIG_0
  6895   6897        parseThat syscall.ptrace.return     0
  6895   6897        parseThat syscall.ptrace            PTRACE_CONT, 6901, 0x0, SIG_0
  6895   6897        parseThat syscall.ptrace.return     0
  6895   6898        parseThat syscall.wait4             -1, 0x7fdac3ffed1c, __WALL, 0x0
  6899   6901      segv-thread signal.send               SIGSEGV -> 6899 6901 segv-thread
  6899   6901      segv-thread signal.send               SIGCHLD -> 6895 6897 parseThat
  6895   6898        parseThat syscall.wait4.return      6901 (WSTOPSIG=SIGSEGV)
  6895   6898        parseThat syscall.wait4             -1, 0x7fdac3ffed1c, __WALL, 0x0
  6895   6895        parseThat signal.send               SIGSTOP -> 6899 6899 segv-thread
  6899   6899      segv-thread signal.send               SIGCHLD -> 6895 6897 parseThat
  6895   6898        parseThat syscall.wait4.return      6899 (WSTOPSIG=SIGSTOP)
  6895   6897        parseThat syscall.ptrace            PTRACE_CONT, 6899, 0x0, SIG_0
  6895   6897        parseThat syscall.ptrace.return     0
  6895   6897        parseThat syscall.ptrace            PTRACE_CONT, 6901, 0x0, SIGSEGV
  6895   6897        parseThat syscall.ptrace.return     0
  6899   6901      segv-thread signal.handle             SIGSEGV
  6895   6898        parseThat syscall.wait4             -1, 0x7fdac3ffed1c, __WALL, 0x0
  6899   6899      segv-thread signal.handle             SIGKILL
  6899   6899      segv-thread kprocess.exit             code=11
  6899   6899      segv-thread signal.send               SIGCHLD -> 6895 6897 parseThat
  6895   6898        parseThat syscall.wait4.return      6899 (WSTOPSIG=SIGTRAP | PTRACE_EVENT_EXIT << 8)
  6895   6898        parseThat syscall.wait4             -1, 0x7fdac3ffed1c, __WALL, 0x0
  6895   6895        parseThat signal.send               SIGSTOP -> 6899 6901 segv-thread
```

Here it hangs.  I'm guessing it's hoping for 6901 to report in another wait return?

For completeness, here's what happens after I hit Ctrl-C:

```
 15986  15986    kworker/u16:1 signal.send               SIGINT -> 6899 6899 segv-thread
 15986  15986    kworker/u16:1 signal.send               SIGINT -> 6895 6895 parseThat
  6895   6895        parseThat signal.handle             SIGKILL
  6895   6895        parseThat kprocess.exit             code=2
  6895   6897        parseThat signal.handle             SIGKILL
  6895   6897        parseThat kprocess.exit             code=2
  6895   6898        parseThat syscall.wait4.return      -512
  6895   6896        parseThat signal.handle             SIGKILL
  6895   6898        parseThat signal.handle             SIGKILL
  6895   6896        parseThat kprocess.exit             code=2
  6895   6898        parseThat kprocess.exit             code=2
  6895   6897        parseThat kprocess.release          released 6895 6897 parseThat
  6895   6896        parseThat kprocess.release          released 6895 6896 parseThat
  6895   6898        parseThat kprocess.release          released 6895 6898 parseThat
 18449  18449             bash kprocess.release          released 6895 6895 parseThat
  6899   6901      segv-thread kprocess.exit             code=139
  6899   6901      segv-thread kprocess.release          released 6899 6901 segv-thread
  2037   2037          systemd kprocess.release          released 6899 6899 segv-thread
```

So the only further action on 6901 is to actually exit from that SIGSEGV.

---

Running on it's own just looks like this:

```
  7137   7137      segv-thread kprocess.exec_complete    
  7137   7138      segv-thread kprocess.start            
  7137   7138      segv-thread signal.send               SIGSEGV -> 7137 7138 segv-thread
  7137   7138      segv-thread signal.handle             SIGSEGV
  7137   7137      segv-thread signal.handle             SIGKILL
  7137   7137      segv-thread kprocess.exit             code=11
  7137   7138      segv-thread kprocess.exit             code=139
  7137   7138      segv-thread kprocess.release          released 7137 7138 segv-thread
 18449  18449             bash kprocess.release          released 7137 7137 segv-thread
```

**Comment by wrwilliams** [comment]
So the threads are still hanging out but gdb doesn't believe they are (for me, at least). 6895-6898 should be the main, ptracer, generator, and handler thread in I believe that order ordinarily, and they've only seen a sigstop if I'm reading this right. And I presume that the release calls are continues.

Do you get a 4-thread backtrace with gdb (in which case there's something weird with my debugging environment) or a 2-thread one (in which case WTF)?

**Comment by cuviper** [comment]
The `kprocess.release` is when the task is entirely dead and getting reaped, following the `kprocess.exit` that merely made it a zombie.  So for instance, 6899 exited a few lines before the hang, but it wasn't released until the very end of that trace - reaped by systemd since I had already killed the actual parent.

I get all 4 threads in gdb:

```
(gdb) info threads
  Id   Target Id         Frame 
* 1    Thread 0x7ffff7fb0740 (LWP 9835) "parseThat" 0x00007ffff50debc0 in pthread_cond_wait@@GLIBC_2.3.2 ()
   from /lib64/libpthread.so.0
  2    Thread 0x7ffff3b7e700 (LWP 9840) "parseThat" 0x00007ffff50e1add in __lll_lock_wait ()
   from /lib64/libpthread.so.0
  3    Thread 0x7ffff337d700 (LWP 9841) "parseThat" 0x00007ffff50debc0 in pthread_cond_wait@@GLIBC_2.3.2 ()
   from /lib64/libpthread.so.0
  4    Thread 0x7ffff2974700 (LWP 9842) "parseThat" 0x00007ffff6e00f0b in waitpid () from /lib64/libc.so.6
(gdb) thread apply all backtrace

Thread 4 (Thread 0x7ffff2974700 (LWP 9842)):
#0  0x00007ffff6e00f0b in waitpid () from /lib64/libc.so.6
#1  0x00007ffff6846384 in GeneratorLinux::getEvent (this=0x6a3780, block=<optimized out>) at /home/jistone/paradyn/dyninst/proccontrol/src/linux.C:158
#2  0x00007ffff678d9b6 in Dyninst::ProcControlAPI::Generator::getMultiEvent (this=0x6a3780, block=<optimized out>, events=std::vector of length 0, capacity 0) at /home/jistone/paradyn/dyninst/proccontrol/src/generator.C:166
#3  0x00007ffff678c658 in Dyninst::ProcControlAPI::Generator::getAndQueueEventInt (this=this@entry=0x6a3780, block=block@entry=true) at /home/jistone/paradyn/dyninst/proccontrol/src/generator.C:208
#4  0x00007ffff677d3b5 in Dyninst::ProcControlAPI::GeneratorMT::main (this=this@entry=0x6a3780) at /home/jistone/paradyn/dyninst/proccontrol/src/generator.C:418
#5  0x00007ffff678cdf3 in Dyninst::ProcControlAPI::GeneratorMT::start (this=0x6a3780) at /home/jistone/paradyn/dyninst/proccontrol/src/generator.C:410
#6  0x00007ffff685b06c in thread_init (d=<optimized out>) at /home/jistone/paradyn/dyninst/common/src/dthread-unix.C:56
#7  0x00007ffff50d95ba in start_thread () from /lib64/libpthread.so.0
#8  0x00007ffff6e3b7cd in clone () from /lib64/libc.so.6

Thread 3 (Thread 0x7ffff337d700 (LWP 9841)):
#0  0x00007ffff50debc0 in pthread_cond_wait@@GLIBC_2.3.2 () from /lib64/libpthread.so.0
#1  0x00007ffff677e02a in boost::condition_variable_any::wait<Mutex<false> > (m=..., this=0x698a20) at /usr/include/boost/thread/pthread/condition_variable.hpp:180
#2  CondVar<Mutex<false> >::wait (this=0x698a20) at /home/jistone/paradyn/dyninst/common/src/dthread.h:121
#3  LinuxPtrace::main (this=this@entry=0x698960) at /home/jistone/paradyn/dyninst/proccontrol/src/linux.C:3342
#4  0x00007ffff6847279 in start_ptrace (lp=lp@entry=0x698960) at /home/jistone/paradyn/dyninst/proccontrol/src/linux.C:3324
#5  0x00007ffff685b06c in thread_init (d=<optimized out>) at /home/jistone/paradyn/dyninst/common/src/dthread-unix.C:56
#6  0x00007ffff50d95ba in start_thread () from /lib64/libpthread.so.0
#7  0x00007ffff6e3b7cd in clone () from /lib64/libc.so.6

Thread 2 (Thread 0x7ffff3b7e700 (LWP 9840)):
#0  0x00007ffff50e1add in __lll_lock_wait () from /lib64/libpthread.so.0
#1  0x00007ffff50dba66 in pthread_mutex_lock () from /lib64/libpthread.so.0
#2  0x00007ffff67cf9f5 in boost::recursive_mutex::lock (this=0x67dde8) at /usr/include/boost/thread/pthread/recursive_mutex.hpp:113
#3  MTManager::startWork (this=0x67dd70) at /home/jistone/paradyn/dyninst/proccontrol/src/process.C:8642
#4  MTManager::evhandler_main (this=0x67dd70) at /home/jistone/paradyn/dyninst/proccontrol/src/process.C:8585
#5  0x00007ffff685b06c in thread_init (d=<optimized out>) at /home/jistone/paradyn/dyninst/common/src/dthread-unix.C:56
#6  0x00007ffff50d95ba in start_thread () from /lib64/libpthread.so.0
#7  0x00007ffff6e3b7cd in clone () from /lib64/libc.so.6

Thread 1 (Thread 0x7ffff7fb0740 (LWP 9835)):
#0  0x00007ffff50debc0 in pthread_cond_wait@@GLIBC_2.3.2 () from /lib64/libpthread.so.0
#1  0x00007ffff67a2fa0 in boost::condition_variable_any::wait<Mutex<false> > (m=..., this=0x7fffec0009b8) at /usr/include/boost/thread/pthread/condition_variable.hpp:180
#2  CondVar<Mutex<false> >::wait (this=0x7fffec0009b8) at /home/jistone/paradyn/dyninst/common/src/dthread.h:121
#3  MailboxMT::dequeue (this=0x7fffec0008c0, block=true) at /home/jistone/paradyn/dyninst/proccontrol/src/mailbox.C:156
#4  0x00007ffff67cc1d9 in int_process::waitAndHandleEvents (block=block@entry=true) at /home/jistone/paradyn/dyninst/proccontrol/src/process.C:1038
#5  0x00007ffff67cf344 in Dyninst::ProcControlAPI::Process::handleEvents (block=block@entry=true) at /home/jistone/paradyn/dyninst/proccontrol/src/process.C:6179
#6  0x00007ffff7ada402 in PCEventMuxer::wait_internal (this=this@entry=0x7ffff7dd7e80 <PCEventMuxer::muxer_>, block=block@entry=true) at /home/jistone/paradyn/dyninst/dyninstAPI/src/pcEventMuxer.C:101
#7  0x00007ffff7ada550 in PCEventMuxer::wait (block=block@entry=true) at /home/jistone/paradyn/dyninst/dyninstAPI/src/pcEventMuxer.C:74
#8  0x00007ffff7a0010c in BPatch::waitForStatusChange (this=<optimized out>) at /home/jistone/paradyn/dyninst/dyninstAPI/src/BPatch.C:1354
#9  0x00000000004109ad in launch_mutator () at /home/jistone/paradyn/dyninst/parseThat/src/dyninstCore.C:584
#10 0x00000000004051d3 in runParseThat (bannerLen=<optimized out>) at /home/jistone/paradyn/dyninst/parseThat/src/parseThat.C:91
#11 0x000000000040490a in main (argc=<optimized out>, argv=<optimized out>) at /home/jistone/paradyn/dyninst/parseThat/src/parseThat.C:77
```

**Comment by wrwilliams** [comment]
@mxz297 and I have been on the trail of other proccontrol bugs this afternoon, mostly Windows-specific, but there may be some overlap here. This certainly looks like our standard starvation because we forgot to continue...something...that's the common proccontrol failure mode.

**Comment by wrwilliams** [comment]
Fun data point here: Xiaozhu built a C++ version of the test mutatee and was not able to reproduce this (including on my development box). Does that suggest any particular behavioral differences to anyone?

**Comment by cuviper** [comment]
I don't see any behavioral difference between compiling with `gcc` or `g++`.
(my command line: `g++ -pthread segv-thread.c -o segv-thread`)

**Comment by wrwilliams** [comment]
The obvious binary level difference seems to be the presence of [de]register_tm_clones in the mutatee that works; we're still trying to nail down how the build differs. Could be timing changes. Any lock that's held when a thread crashes should be released by the kernel, right?

**Comment by cuviper** [comment]
The kernel does not release locks.  If it was a robust mutex (not the default) then the kernel will mark it with `FUTEX_OWNER_DIED`, and the next attempt to lock the same will get `EOWNERDEAD`.  The lock can be cleaned up with `pthread_mutex_consistent()`.

**Comment by wrwilliams** [comment]
My build line was gcc segv-thread.c -o segv-thread -lpthread, which should be equivalent to yours, I think.

I'm really suspecting that there's some hidden mutatee-side deadlock going on, FWIW.

**Comment by cuviper** [comment]
The thread died as soon as it got that `PTRACE_CONT` with `SIGSEGV`, and the main thread got a `SIGKILL` right after, because one death kills everything.  So what are you expecting the mutatee to be doing?

**Comment by wrwilliams** [comment]
I had been expecting the main thread to be hanging out waiting for a lock (e.g. in exit instrumentation) that it never would acquire, but SIGKILL really should put paid to that notion, it's true.

**Comment by cuviper** [comment]
If I hack out the option `PTRACE_O_TRACEEXIT`, then everything finishes properly.

If I use gdb in the ptrace thread to inject a `PTRACE_CONT,0` on the thread that had reported `PTRACE_EVENT_EXIT`, then this also finishes properly.  This seems consistent with the manpage, which states:

>        PTRACE_EVENT_EXIT
>               Stop before exit (including death from exit_group(2)), signal
>               death, or exit caused by execve(2) in a multithreaded process.
>               PTRACE_GETEVENTMSG returns the exit status.  Registers can be
>               examined (unlike when "real" exit happens).  The tracee is
>               still alive; it needs to be PTRACE_CONTed or PTRACE_DETACHed
>               to finish exiting.

As near as I can make out, what happens here is:
- sub thread raises `SIGSEGV`
- the mutator is notified and delivers the `SIGSEGV` with `PTRACE_CONT`
- the sub thread's death sends a `SIGKILL` to the main thread
- the mutator is notified of the main thread's `PTHREAD_EVENT_EXIT`
  - the main thread remains stuck in this ptrace stop
- the mutator tries to stop all threads in `syncRunState`, sends `SIGSTOP` to the sub thread
  - but that sub thread is already too dead to care
- the mutator waits, but one child is dead and the other already stopped...

**Comment by wrwilliams** [comment]
syncRunState should IIRC only be trying to stop everything via SIGSTOP if we're trying to deliver an exit callback. Exit callbacks are *not* normal callbacks re: what can actually be done from them; is PTRACE_EVENT_EXIT a sufficient guarantee that all threads of the tracee are either dead or signal-delivery-stopped such that we can deliver a callback safely in this context?

**Comment by cuviper** [comment]
Between the "Death under ptrace" section of the manpage and the notes in the "BUGS" section about `SIGKILL`, I don't feel very comfortable making any promises about the order here.  But in general, no, a `PTRACE_EVENT_EXIT` could simply be one thread's normal exit while the process continues.

It looks like `SIGSEGV` is sending that thread into the kernel's `do_coredump()`, which calls `coredump_wait()`, and this is where it kills the other threads and waits for them to exit.  The thread won't continue to its own `do_exit()` (where it sends `PTRACE_EVENT_EXIT`) until everyone else is gone.

So it doesn't seem like it will be reliable to try stopping the other threads for an exit callback, because any of those threads might be stuck waiting for this one to complete.  ("Stuck waiting" in the kernel sense, where there's nothing userspace can do to break it out until its wait conditions are met.)

**Comment by wrwilliams** [comment]
Hrm. I thought we could at least be sure that `PTRACE_EVENT_EXIT` on the main LWP was going to be a whole-process exit; certainly our decoder logic attempts to map `PTRACE_EVENT_EXIT` to the appropriate thread/process level exit decision. And I don't think we can get into the bad state here without a user-level process exit callback present (need to verify, but that should be a trivial change to parseThat).

Two questions spring to mind:

1) Is there an error that we should be getting from our attempt to send a `SIGSTOP` to the thread that's blocked in `do_coredump`, but are failing to process?
2) Is there any good way to introduce a timeout into our ptracer such that it can report "failed to get expected event" when we're expecting a `SIGSTOP` (or some other proccontrol-generated signal) and the kernel may keep us from receiving it?

**Comment by cuviper** [comment]
I tried another experiment where the thread calls `_exit(0)` instead, so there's no coredump interference.  Both threads generate `PTHREAD_EVENT_EXIT`, but it is not consistent in the order these are delivered.  Most of the times I tried, the main thread reported first, and everyone was happy.  One time they were reversed though, and I saw one thread try to send that `SIGSTOP` at the same the other thread was receiving its exit -- this case also confused and hung the mutator.

> 1) Is there an error that we should be getting from our attempt to send a SIGSTOP to the thread that's blocked in do_coredump, but are failing to process?

No, the `tkill` is succeeding.  I suppose that's because the thread is not dead-dead, only mostly dead.

> 2) Is there any good way to introduce a timeout into our ptracer such that it can report "failed to get expected event" when we're expecting a SIGSTOP (or some other proccontrol-generated signal) and the kernel may keep us from receiving it?

I guess what you mean is we're expecting a ptrace-stop for the child's `SIGSTOP`, right?  I don't know a good way to timeout `waitpid`, but you can use `WNOHANG` paired with `sigtimedwait()` for `SIGCHLD`, if you make sure that signal is only delivered to the waiting thread.

Or the other threads that are waiting on a condvar for those events could use a `timed_wait`.

**Comment by wrwilliams** [comment]
> I guess what you mean is we're expecting a ptrace-stop for the child's SIGSTOP, right? I don't know a good way to timeout waitpid, but you can use WNOHANG paired with sigtimedwait() for SIGCHLD, if you make sure that signal is only delivered to the waiting thread.
> Or the other threads that are waiting on a condvar for those events could use a timed_wait.

We're expecting, at a minimum, that there will be some ptrace event for the child after we send it a SIGSTOP if we haven't already seen its LWP exit. That may be the stop, it may be LWP death, it may be a different signal, but we expect *something* coming back from ptrace (or I suppose already in the ptrace event queue). It appears that expectation is not a valid one, though, and that the LWP exit event may not be delivered at all (or in the sequence we expect).

Are we getting the SIGCHLD for the subthread exit before we get the SIGKILL on the main thread? That's the only way I can see that the SIGCHLD helps us, is if we're guaranteed to see those events in that order...

**Comment by cuviper** [comment]
> the LWP exit event may not be delivered at all (or in the sequence we expect).

The latter -- that LWP exit is coming, just not guaranteed in any order.  Although in the manpage it also sounds like `SIGKILL` can't be promised to generate an exit either, even though it does for now.  Actually, life here might be easier if it *didn't*, because then `coredump_wait` wouldn't get stuck on those ptrace events for other threads.  (You'd just get `ESRCH` for those killed.)  Alas.

> Are we getting the SIGCHLD for the subthread exit before we get the SIGKILL on the main thread?

No, the `SIGCHLD` isn't generated until the exit event is actually ready to be retrieved.  I only mentioned it because watching that signal lets you use a `WNOHANG` wait and do your blocking with timeouts elsewhere.

**Comment by wrwilliams** [comment]
I have a simple and elegant solution, which may also be wrong...

We should get one LWP exit per LWP in the process, regardless of the order in which we receive them. Is there any good reason not to replace this:

```                    pthrd_printf("Decoded event to pre-exit on %d/%d\n",
                                  proc->getPid(), thread->getLWP());
                    if (thread->getLWP() == proc->getPid())
		            {
```

with this:

```                    pthrd_printf("Decoded event to pre-exit on %d/%d\n",
                                  proc->getPid(), thread->getLWP());
                    if (proc->threadPool()->size <= 1)
		            {
```
in the decoder, delaying process exit until we're exiting our last LWP (in whatever order)?

**Comment by cuviper** [comment]
What interaction is the user expecting to be possible in this callback?  If it's just reading memory, then the particular thread doesn't matter.  If they want to read registers or TLS, then the thread matters, but I don't know that we can make any promises.

It sounds like with `SIGKILL` we're not *guaranteed* to see every `PTRACE_EVENT_EXIT`, but it does work that way now. (`do_exit` is called for every LWP, and this unconditionally calls `ptrace_event`)

Also, "execve(2) under ptrace" is interesting.  All other threads get `PTRACE_EVENT_EXIT`, and the exec thread resets its thread ID to match the process ID.  So your new suggestion probably will behave better when a thread execs, where again the main thread exit isn't a total process exit.  (The reset thread ID might be confusing in other places though.)

**Comment by wrwilliams** [comment]
AFAIK the user should be able to read memory and get the process exit code.

Unfortunately, at least on RHEL6 the above is simple, obvious, and wrong--we convert the main thread's death to a user thread death followed by an lwp death. However, we've got callbacks for those as well, and they attempt to stop the main thread (okay) and the subthread that's SEGVing (not okay).

Log at the point of hang:

[process.C:835-U] - Checking if any ProcStop events on 6114 are ready
[process.C:3285-U] - 6114/6114: setting target state to Stopped
[process.C:3285-U] - 6114/6115: setting target state to Stopped
[process.C:847-U] - Current Threading State for 6114:
[process.C:863-U] - 6114/6114 hand: stopped, gen: stopped, user: stopped, code: ----------------SSS
[process.C:863-U] - 6114/6115 hand: running, gen: running, user: stopped, code: ----------------SRR
[process.C:872-U] - Running plat_syncRunState on 6114
[process.C:2911-U] - intStop on thread 6114/6115
[process.C:3107-U] - Setting pending stop to true, thread 6114/6115
[process.C:8342-U] - Adjusting counter PendingStops by 1, before 0, after 1
[process.C:4289-U] - Desyncing 6114/6115 pending stop state to level 1
[process.C:4411-U] - Changing pending stop state for 6114/6115 from dontcare to running
[process.C:3126-U] - 	 Pending stop level is 1
[linux.C:1590-U] - Sending 19 to 6115
[process.C:1024-U] - Blocking for events = !F && ((F && !? && ?) || F || T || ? || ? || ? || ? || ? || ?)
[mailbox.C:155-U] - Blocking for events from mailbox, queue size = 0

Is there any way to work around this? I'm leaning towards a wontfix for 9.2, because this is starting to open some pretty big cans of worms in callback design.

**Comment by cuviper** [comment]
I wasn't sure if every thread will exit with the true process exit code, but it appears that they do.

I'm not sure of any workaround here.  Letting the user take any control during a `PTRACE_EVENT_EXIT` seems prone to this sort of stuck state.  Perhaps we should use a breakpoint form of pre-exit for normal exits, and forget about catching signaled exits?  (You'd still get the normal WIFSIGNALED wait status.)

**Comment by cuviper** [comment]
At the Scalable Tools Workshop we had a session discussing this bug.  We thought about how we might have to restrict what's allowed in some event callbacks, but that was cutting out use-cases that are generally fine.  We instead decided to look into waitpid timeouts (w/ something like `sigtimedwait` and a `WNOHANG` loop) if we could find a reliable indication that a thread is in a "bad" state.  I wrote in a followup email:

> The state character of the do_coredump thread is 'D', uninterruptible
> sleep (usually IO), so that's not too helpful.
> 
> However, the "flags" field of /proc/PID/stat might do the trick.  This
> problem thread has PF_DUMPCORE, so that seems a pretty clear sign.  I
> think PF_EXITING and PF_SIGNALED are other good "can't stop this"
> candidates, although those should be on their way to proper exit
> notifications -- I guess that's unstoppable anyway.
> 
> But we can't just check for those right away when trying to stop a
> thread.  It may still be in a perfectly good state when we try to
> SIGSTOP, then receive a SIGSEGV/whatever before our stop is actually
> delivered.  So I think we'll still need a timeout when trying to stop
> any thread, then look at those flags to decide if we think it will ever
> stop.  If we don't see a "bad" PF flag, do nothing and wait+timeout
> again, as often as necessary.  So I think we can get away with a
> relatively short timeout for good user experience, say 1 second, without
> causing problems for normal cases that are just slow.

**Comment by cuviper** [comment]
If we have to mask `SIGCHLD` before waiting on it, things will get hairy.  If nothing else, signal masks are inherited across `fork` and preserved across `execve`, so we need to be careful about affecting child processes.  Especially since this is a library, we don't control our whole process.

Another timeout implementation possibility is with `signalfd`, then use a `select`/`poll`/etc with a timeout.  It's not clear to me if this really needs to mask `SIGCHLD` though.

It might be that the timeouts need to go elsewhere, and leave the current blocking `waitpid` alone.

**Comment by mxz297** [comment]
Will check if the same problem still exists in Dyninst 10.1

**Comment by mxz297** [comment]
No longer seen in Dyninst 10.1
